### PR TITLE
Metric name validation

### LINKED ISSUE #2586: Remote read should validate returned labelnames
While playing around I managed to create a labelname called `12a`, which is clearly invalid.

We should validate the returned labelnames, and reject results with invalid ones.

### PR DESCRIPTION
**Pull request by Conorbro**
Fixes #2586

### CODE REVIEW

File: storage/remote/read_test.go [hunk 1]
```diff
@@ -22,6 +23,45 @@ import (
 	"github.com/prometheus/prometheus/storage/metric"
 )
 
+func TestValidateNameAndLabels(t *testing.T) {
+
```
Comment (grobie) [inline storage/remote/read_test.go]: Nitpick: It's not common in stdlib or Prometheus golang code to include empty lines at the beginning or end of functions.

File: storage/remote/read.go [hunk 2]
```diff
@@ -95,9 +96,26 @@ func (q *querier) read(ctx context.Context, from, through model.Time, matchers m
...
+
 	return res, err
 }
 
+// validateNamesAndValues rejects invalid metric/label names returned from remote read.
+func validateNamesAndValues(res model.Matrix) {
+	for i, r := range res {
+		for name, value := range r.Metric {
+			if !model.IsValidMetricName(model.LabelValue(name)) || !utf8.ValidString(string(value)) {
+				delete(r.Metric, name)
```
Comment (brian-brazil) [inline storage/remote/read.go]: We should return an error in this case, not just remove the bad label.

File: storage/remote/read.go [hunk 3]
```diff
@@ -95,9 +96,26 @@ func (q *querier) read(ctx context.Context, from, through model.Time, matchers m
...
+	validateNamesAndValues(res)
+
 	return res, err
 }
 
+// validateNamesAndValues rejects invalid metric/label names returned from remote read.
+func validateNamesAndValues(res model.Matrix) {
+	for i, r := range res {
+		for name, value := range r.Metric {
+			if !model.IsValidMetricName(model.LabelValue(name)) || !utf8.ValidString(string(value)) {
```
Comment (brian-brazil) [inline storage/remote/read.go]: The rules for label names are stricter than for metric names, you'll need to check for valid label names.

**Comment by Conorbro** [comment]
@brian-brazil

File: storage/remote/read.go [hunk 4]
```diff
@@ -93,11 +96,29 @@ func (q *querier) read(ctx context.Context, from, through model.Time, matchers m
 	if err != nil {
 		return nil, err
 	}
-
 	removeLabels(res, added)
+	validateNamesAndValues(res)
```
Comment (brian-brazil) [inline storage/remote/read.go]: You need to check the error here

File: storage/remote/read.go [hunk 5]
```diff
@@ -93,11 +96,29 @@ func (q *querier) read(ctx context.Context, from, through model.Time, matchers m
...
+
 	return res, err
 }
 
+// validateNamesAndValues rejects invalid metric/label names returned from remote read.
+func validateNamesAndValues(res model.Matrix) (err error) {
+	for i, r := range res {
+		for name, value := range r.Metric {
+			if !strutil.IsValidLabelName(string(name)) || !utf8.ValidString(string(value)) || !model.IsValidMetricName(value) {
+				delete(r.Metric, name)
```
Comment (brian-brazil) [inline storage/remote/read.go]: You don't need to change the input data, just return an error

File: storage/remote/read.go [hunk 6]
```diff
@@ -93,11 +96,29 @@ func (q *querier) read(ctx context.Context, from, through model.Time, matchers m
...
+				res[i] = &model.SampleStream{
+					Metric: r.Metric,
+					Values: r.Values,
+				}
+				err = errors.New("Invalid label set found")
```
Comment (brian-brazil) [inline storage/remote/read.go]: It'd be useful for debugging to mention which data is bad.

File: storage/remote/read.go [hunk 7]
```diff
@@ -93,11 +96,29 @@ func (q *querier) read(ctx context.Context, from, through model.Time, matchers m
...
+	validateNamesAndValues(res)
+
 	return res, err
 }
 
+// validateNamesAndValues rejects invalid metric/label names returned from remote read.
+func validateNamesAndValues(res model.Matrix) (err error) {
+	for i, r := range res {
+		for name, value := range r.Metric {
+			if !strutil.IsValidLabelName(string(name)) || !utf8.ValidString(string(value)) || !model.IsValidMetricName(value) {
```
Comment (brian-brazil) [inline storage/remote/read.go]: model.IsValidMetricName should only be applied to the __name__ label, not all labels.

File: util/strutil/strconv.go [hunk 8]
```diff
@@ -42,3 +44,11 @@ func GraphLinkForExpression(expr string) string {
 func SanitizeLabelName(name string) string {
 	return invalidLabelCharRE.ReplaceAllString(name, "_")
 }
+
+// IsValidLabelName evaluates whether a given label name is valid or not.
+func IsValidLabelName(name string) bool {
```
Comment (Gouthamve) [inline util/strutil/strconv.go]: Can we reuse this function in common/model to do the same? http://godoc.org/github.com/prometheus/common/model#LabelName.IsValid

Also for metric name validation this one: http://godoc.org/github.com/prometheus/common/model#IsValidMetricName

**Comment by Conorbro** [comment]
@brian-brazil

File: storage/remote/read_test.go [hunk 9]
```diff
@@ -22,6 +23,115 @@ import (
...
+			expectedErr: "Invalid metric name: @invalid_name",
+			shouldPass:  false,
+		},
+	}
+
+	for _, test := range tests {
+		err := validateLabelsAndMetricName(test.result)
+		if test.shouldPass {
+			if err != nil {
+				log.Fatalf("Test should pass, got unexpected error: %v", err)
```
Comment (brian-brazil) [inline storage/remote/read_test.go]: use t.Fatalf, not log.Fatalf

**Comment by brian-brazil** [comment]
Thanks!

**Comment by tomwilkie** [comment]
This will need porting to v2 also

**Pull request merged**